### PR TITLE
Resolver improvements

### LINKED ISSUE #231: NXDOMAIN reported randomly for domains being browsed
**Checklist**:

- [ ] I'm using the official portmaster release (i.e no custom builds)
- [ ] I searched for similar/existing issues first.
- [ ] My issue is not mentioned in the Known Issues section of my [OS](https://github.com/safing/portmaster/wiki)

**What happened**:

All pages being browsed returns NXDOMAIN error randomly. It happens more and more and after there is "calm period" which can be even a day. Today I was workin in internet banking and I got 1 error every ~2 page loads.

When reporting this I even noticed the SERVFAIL from privacytools.

```
Jan 12 21:25:03 northys-fedora portmaster-start[61220]: 210112 21:25:03.732 nameserver:281 ▶ INFO 596 nameserver: returning accepted response (SERVFAIL) for bin.privacytools.io.AAAA to northys:/usr/lib64/firefox/firefox:57338
Jan 12 21:25:05 northys-fedora portmaster-start[61220]: 210112 21:25:05.228 nameserver:281 ▶ INFO 606 nameserver: returning accepted response (SERVFAIL) for bin.privacytools.io.AAAA to northys:/usr/lib64/firefox/firefox:57338
```

**What you expected to happen**:

No DNS errors at all.

**How to reproduce it (as minimally and precisely as possible)**:

```
{
  "core": {
    "expertiseLevel": "developer"
  },
  "dns": {
    "nameservers": [
      "dot://1.0.0.2:853?verify=cloudflare-dns.com\u0026name=Cloudflare\u0026blockedif=zeroip",
      "dot://94.130.106.88:443?verify=dot1.applied-privacy.net\u0026name=AppliedPrivacy"
    ]
  },
  "filter": {
    "includeCNAMEs": 7,
    "lists": [
      "TRAC",
      "MDLC",
      "17-MAL",
      "BAD"
    ],
    "removeBlockedDNS": 7,
    "removeOutOfScopeDNS": 6
  }
}
```

**Anything else we need to know?**:

**Environment**:

Portmaster Version: 0.6.4 (lol it's not alpha anymore? cool!)

Operating System:
- [ ] Windows 7
- [ ] Windows 8/8.1
- [ ] Windows 10
- [x] Linux

If applicable you can provide related sections from the log files and ensure to **remove sensitive or otherwise private information**.

https://bin.privacytools.io/?d43c77a94a7fe61a#BAjFrNVVZzivJpa9YY3K16RVdhR3zrC1LbZhMMKLTjwQ

### PR DESCRIPTION
**Pull request by dhaavi**
0e93204 might fix #231

### CODE REVIEW

File: resolver/api.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	resolversLock.RLock()
+	defer resolversLock.RUnlock()
+
+	export := make([]*resolverExport, 0, len(globalResolvers))
+	for _, r := range globalResolvers {
+		export = append(export, &resolverExport{
+			Resolver: r,
+			Failing:  r.Conn.IsFailing(),
+		})
+	}
```
Comment (ppacher) [inline resolver/api.go:46]: There's no reason to use `[]*resolverExport`, use `[]resolverExport` to save one level of indirection here.

File: resolver/namerecord.go [hunk 2]
```diff
@@ -88,6 +88,10 @@ func GetNameRecord(domain, question string) (*NameRecord, error) {
 
 // DeleteNameRecord deletes a NameRecord from the database.
 func DeleteNameRecord(domain, question string) error {
+	// In order to properly delete an entry, we must also clear the caches.
+	recordDatabase.FlushCache()
+	recordDatabase.ClearCache()
+
 	key := makeNameRecordKey(domain, question)
 	return recordDatabase.Delete(key)
 }
```
Comment (ppacher) [inline resolver/namerecord.go:97]: This get's called everytime a rrCache record is expired and we are going to clear the whole cache everytime this happens. We should check how much performance we are going to lose here if we clear the cache and start refilling it over and over again.

File: resolver/scopes.go [hunk 3]
```diff
@@ -208,10 +208,6 @@ func (q *Query) checkCompliance() error {
 }
 
 func (resolver *Resolver) checkCompliance(_ context.Context, q *Query) error {
-	if q.FQDN == resolver.SkipFQDN {
-		return errSkip
-	}
-
```
Comment (ppacher) [inline resolver/scopes.go:214]: `errSkip` is not used anymore but still defined.

File: resolver/namerecord.go (hunk 2 above)
Comment (dhaavi) [reply resolver/namerecord.go:97]: This function is only used for resetting the cache for a single domain/query - I improved the name to better reflect that.

**Pull request merged**